x-pack/plugin/shutdown/src/test/java/org/elasticsearch/xpack/shutdown/NodeSeenServiceTests.java

```diff
@@ -0,0 +1,236 @@
+        return masterServiceTaskQueue;
+    }
+
+    public void testCleanIfRemoved() {
+        ClusterService clusterService = mock(ClusterService.class);
+        final ThreadPool mockThreadPool = mock(ThreadPool.class);
+        when(mockThreadPool.absoluteTimeInMillis()).thenReturn(120_000L);
+        when(clusterService.threadPool()).thenReturn(mockThreadPool);
+        MasterServiceTaskQueue<RemoveSigtermShutdownTask> taskQueue = newMockTaskQueue(clusterService);
+        NodeSeenService nodeSeenService = new NodeSeenService(clusterService);
+        var master = TestDiscoveryNode.create("node1", randomNodeId());
+        var other = TestDiscoveryNode.create("node2", randomNodeId());
+        var another = TestDiscoveryNode.create("node3", randomNodeId());
+
+        nodeSeenService.clusterChanged(
+            new ClusterChangedEvent(
+                this.getTestName(),
+                createClusterState(new NodesShutdownMetadata(Map.of(another.getId(), createShutdown(another.getId(), 0))), master, other),
+                createClusterState(null, master, other, another)
+            )
+        );
+
+        Mockito.verify(taskQueue, times(1))
+            .submitTask(eq("sigterm nodes left cluster"), any(NodeSeenService.RemoveSigtermShutdownTask.class), isNull());
+    }
+
+    public void testDontCleanIfPresent() {
+        ClusterService clusterService = mock(ClusterService.class);
+        final ThreadPool mockThreadPool = mock(ThreadPool.class);
+        when(mockThreadPool.absoluteTimeInMillis()).thenReturn(120_000L);
+        when(clusterService.threadPool()).thenReturn(mockThreadPool);
+        MasterServiceTaskQueue<RemoveSigtermShutdownTask> taskQueue = newMockTaskQueue(clusterService);
+        NodeSeenService nodeSeenService = new NodeSeenService(clusterService);
+        var master = TestDiscoveryNode.create("node1", randomNodeId());
+        var other = TestDiscoveryNode.create("node2", randomNodeId());
+        var another = TestDiscoveryNode.create("node3", randomNodeId());
+
+        nodeSeenService.clusterChanged(
+            new ClusterChangedEvent(
+                this.getTestName(),
+                createClusterState(
+                    new NodesShutdownMetadata(
+                        Map.of(another.getId(), createShutdown(another.getId(), 0), other.getId(), createShutdown(other.getId(), 20L))
+                    ),
+                    master,
+                    other,
+                    another
+                ),
+                createClusterState(null, master, other, another)
+            )
+        );
+
+        Mockito.verify(taskQueue, never()).submitTask(any(), any(), any());
+    }
+
+    public void testExecutorRemoveStale() {
+        long now = 100_000L;
+        long grace = getDefaultGrace();
+        var master = TestDiscoveryNode.create("master1", randomNodeId());
+        var other = TestDiscoveryNode.create("other2", randomNodeId()); // not term
```

Comment:
just a nit: perhaps you could name these 'notTermNode, (still)existingNode, withinGraceNode, outOfGracenode etc
it would be a little easier to read the assertions. 
but still this is a great test

Reply:
Thanks.  I updated the node names.